feature/text_based.py

```diff
@@ -0,0 +1,702 @@
+            ("random", True, "unicost"): lambda: None,
+            ("greedy", False, "unicost"): lambda: None,
+            ("greedy", True, "unicost"): lambda: None,
+            ("exact", True, "unicost"): lambda: None,
+            ("exact", False, "unicost"): lambda: None
+        }
+
+        options = option_map.get((optimization_method, selection_size is not None, cost_metric))
+        if options:
+            options()
+        else:
+            feature_column = self.featurization_method.fit_transform(input_df)
+            self.features = np.array([eval(l) if isinstance(l, str) else l for l in feature_column.tolist()])
+            check_true(len(self.features) == self._num_cols,
+                       f"Process Data Error: features size ({len(self.features)}) "
+                       f"should match the number of columns ({self._num_cols})")
+
+        check_true(self.matrix.ndim == 2, "Process Data Error: matrix should 2D")
+        if self.selection_size is not None:
+            check_true(self.selection_size <= self._num_cols,
+                       "Process Data Error: selection_size cannot exceed num columns")
+
+    def _select_multi_level_optimization(self) -> List:
+
+        unicost = np.ones(self._num_cols)
+
+        if self.selection_size is None:
+            if self.cost_metric == "unicost":
+                unicost_selected = self._solve_set_cover_cost_option(unicost)
+                selected = unicost_selected
+            else:
+                unicost_selected = self._solve_set_cover_cost_option(unicost)
+                k_unicost = len(unicost_selected)
+                diversity_cost = self._get_diversity_cost(k_unicost)
+                diversity_selected = self._solve_set_cover_cost_option(diversity_cost)
+                selected = diversity_selected
+        else:
+            if self.cost_metric == "unicost":
+                selected = self._solve_set_cover_cost_option(unicost)
+                cost = unicost
+                k = len(selected)
+            else:
+                unicost_selected = self._solve_set_cover_cost_option(unicost)
+                k_unicost = len(unicost_selected)
+                cost = self._get_diversity_cost(k_unicost)
+                selected = self._solve_set_cover_cost_option(cost)
+                k = len(selected)
+
+            if self.selection_size < k:
+                data = Data(cost=cost, matrix=self.matrix)
+                selected = self._solve_max_cover(data, selected)
+                k = len(selected)
+            else:
+                pass
+
+        # Return solution
+        return selected
+
+    def _get_diversity_cost(self, selected_size: int) -> List[float]:
+        """
+        diversity cost:
+            there is a situation where selected_size is equal to the number of rows in features and all diversity
+             costs become zero. There are two ways to handle this situation:
+                * add dummy cost to diversity cost (DUMMY)
```

Comment:
Looks like the dummy cost block has been commented out / removed?

Reply:
The reason why we need to add NEXT_S_DIS or DUMMY is that there are two alternative options to handle situations where the minimum distance between each two features is zero. This situation may occur if the features are very similar or if there are a limited number of features. When this happens, the diversity cost becomes zero. Therefore, the code adds NEXT_S_DIS to replace the zero value in the distances matrix with the next smallest non-zero distance. This ensures that each feature has a non-zero diversity cost, which helps to select a more diverse set of features even when the minimum distance between each two features is zero. Also, DUMMY may not be a good option (it is @takojunior suggestion) for Greedy since the adjusted cost brings all values between (0, 1). In this case, there is not much difference between uni-cost and diverse cost, since the adjusted cost elements are so similar. We need to decide with which option we want to continue.